storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java

```diff
@@ -681,7 +687,12 @@ private static void writeSnapshot(File file, Map<Long, ProducerStateEntry> entri
 
         try (FileChannel fileChannel = FileChannel.open(file.toPath(), StandardOpenOption.CREATE, StandardOpenOption.WRITE)) {
             fileChannel.write(buffer);
-            fileChannel.force(true);
+        }
+
+        if (scheduler != null) {
+            scheduler.scheduleOnce("flush-producer-snapshot", () -> Utils.flushFileQuietly(file.toPath(), "producer-snapshot"));
+        } else {
+            Utils.flushFileQuietly(file.toPath(), "producer-snapshot");
```

Comment:
After this change, the fileChannel needs to open/close twice under `scheduler == null` case. Maybe we can jsut do it like this:
```
try (FileChannel fileChannel = FileChannel.open(file.toPath(), StandardOpenOption.CREATE, StandardOpenOption.WRITE)) {
            fileChannel.write(buffer);
           if (scheduler == null) {
                // directly flush to disk
                fileChannel.force(true);
           }
        }

        if (scheduler != null) {
            scheduler.scheduleOnce("flush-producer-snapshot", () -> Utils.flushFileQuietly(file.toPath(), "producer-snapshot"));
        }
```

WDYT?